### ISSUE TITLE
How can I sent to CloudWatch is a JSON Payload?

### ISSUE THREAD

**Issue by DavidAmesPup** [Cimpress-MCP/serilog-sinks-awscloudwatch#41]
Hi there,

I'd like to send the log output to Cloudwatch as a Json payload, as the Serilog.Formatting.Json.JsonFormatter does.  Is there a way I can achieve this?  The closest thing I can see is to implement ILogEventRenderer, but I can't see how to return the payload differently to make Cloudwatch think it's JSON.

thanks,
dave

**Comment by wparad** [comment]
Here's an example
```csharp
using System;
using System.IO;
using System.Net;
using Newtonsoft.Json;
using Serilog.Events;
using Serilog.Sinks.AwsCloudWatch;

namespace Logging
{
    public class JsonLogMessageRenderer : ILogEventRenderer
    {
        // the cached host name which is part of every log message
        private readonly string hostname;

        public JsonLogMessageRenderer()
        {
            hostname = Dns.GetHostName();
        }

        public string RenderLogEvent(LogEvent logEvent)
        {
            try
            {
                // try logging according to the preferred template
                using (var writer = new StringWriter())
                {
                    // flatten the message as a simpler option to look at it within the RenderedMessage property of the JSON file
                    logEvent.RenderMessage(writer);

                    // create the message object
                    var message = new {logEvent.MessageTemplate, logEvent.Properties, RenderedMessage = writer.ToString(), Level = logEvent.Level.ToString(), Exception = logEvent.Exception?.ToString(), Hostname = hostname};

                    // serialize the object as JSON
                    return JsonConvert.SerializeObject(message);
                }
            }
            catch (Exception exception)
            {
                try
                {
                    var message = new {RenderedMessage = "Failed to render log message.", MessageTemplate = logEvent.MessageTemplate.Text, Exception = exception.ToString()};
                    return JsonConvert.SerializeObject(message);
                }
                catch (Exception ex)
                {
                    // fallback to just return the fact that we were unable to render the log message
                    return "Unable to render log message. Reason was " + ex;
                }
            }

        }
    }
}
```
and
```csharp
            var options = new CloudWatchSinkOptions
            {
                LogGroupName = logGroupName,
                LogEventRenderer = new JsonLogMessageRenderer()
            };
            var client = new AmazonCloudWatchLogsClient();
            var loggerConfig = new LoggerConfiguration().WriteTo.AmazonCloudWatch(options, client);
```

Does that solve your issue?

**Comment by chrisoverzero** [comment]
What I have done in my applications is implement `ILogEventRenderer` in such a way that I call the built-in `Serilog.Formatting.Json.JsonFormatter` in `ILogEventRenderer.RenderLogEvent`.

It works, but it seems like an unnecessary extra step. Why does this library create `ILogEventRenderer` instead of accepting a `Serilog.Formatting.ITextFormatter`?

**Comment by thoean** [comment]
I thinik it'd be a fine idea to support both, the current version as well as the `ITextFormatter` option. Is that something you want to address in a PR? #22 addressed that a few months back, but unfortunately it got abandoned...

**Comment by wparad** [comment]
I'm almost thinking it should be a permanent addition to the library as long as it doesn't depend on any other libs.  One problem is that the the logEvent doesn't have the input objects to the `Log` call, it has some custom Serilog format.  Which means you'll need to still have some complex logic, see [MessageTemplateTextFormatter.cs](http://sourcebrowser.io/Browse/serilog/serilog/src/Serilog/Formatting/Display/MessageTemplateTextFormatter.cs#56) for the current implementation.

At that point you'll still need to override something, but the `ILogEventRenderer` seems like an easier override than the `ITextFormatter`.  Before you create the PR, I would suggest creating what the TextFormatter would look like, and then we can decide if either/both the override and supporting library code make sense to add in.

Thanks for continuing this work.

**Comment by wparad** [comment]
Closed via #50
